include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1012 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_DECODED_VIEW_HPP
+#define BOOST_URL_DECODED_VIEW_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <boost/url/string_view.hpp>
+#include <boost/url/pct_encoding.hpp>
+#include <type_traits>
+#include <iterator>
+#if !defined(BOOST_NO_CXX17_HDR_STRING_VIEW)
+#include <string_view>
+#endif
+
+namespace boost {
+namespace urls {
+
+/** A view on a string as percent decoded characters
+
+    Objects of this type represent a range of
+    characters decoded from a percent encoded
+    string. The object has no ownership of the
+    underlying character buffer.
+
+    These views are references to the encoded
+    strings that allow cheap copies.
+
+    The correctness of the underlying string
+    is assumed. The values returned by
+    `decoded_view::iterator` when the
+    underlying values are an invalid
+    percent-encoded string is undefined behavior.
+
+    The function @ref validate_pct_encoding
+    can be used to validate the underlying string.
+
+*/
+class decoded_view
+{
+    char const* p_{nullptr};
+    std::size_t n_{0};
+    std::size_t dn_{0};
+    bool plus_to_space_{true};
+
+    using traits_type = std::char_traits<char>;
+
+public:
+    /** Type of a decoded character
+    */
+    using value_type = char;
+
+    /// @copydoc value_type
+    using reference = char;
+
+    /// @copydoc value_type
+    using const_reference = char;
+
+    /** The unsigned integer type used to represent size.
+    */
+    using size_type = std::size_t;
+
+    /** The signed integer type used to represent differences.
+    */
+    using difference_type = std::ptrdiff_t;
+
+    /** A read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+#ifdef BOOST_URL_DOCS
+    using const_iterator = __see_below__;
+#else
+    class const_iterator;
+#endif
+
+    /// @copydoc const_iterator
+    using iterator = const_iterator;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+     */
+    BOOST_CONSTEXPR
+    decoded_view(
+        bool plus_to_space = true) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        bool plus_to_space = true) noexcept
+        : p_(str)
+        , n_(traits_type::length(str))
+        , dn_(pct_decode_bytes_unchecked(string_view(str)))
```

Comment:
User inputs need to be validated, we should not used the unchecked algorithm here